### PR TITLE
Retain focus and keyboard visibility of SearchTextField after configuration changes

### LINKED ISSUE #1033: [FR]: Retain states of SearchTextField after configuration changes
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the problem

Unlike `EditText` in Android Views, `TextField` in Compose doesn’t automatically retain or restore its states as the device goes through configuration changes such as rotating the device between landscape and portrait orientations, folding / unfolding the device, resizing window, switching dark mode, etc.
Currently, the app opens keyboard after configuration changed even when a user hid the keyboard by tapping back button or 'search' key. Also, position of cursor is not retained after the event.

### Describe the solution

I suggest to implement equivalent experiences with EditText, by retaining the following after configuration changes:
- Position of cursor
- Text selection
- Focus
- Keyboard visibility

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by saryongkang**
**What I have done and why**
Currently, the app opens keyboard after configuration changed even when a user hid the keyboard by tapping back button or 'search' key. Also, position of cursor is not retained after the event.
Therefore, I...
- added `focusRestorer` modifier to retain focus and keyboard keyboard 
- added logic to retain focus and keyboard after configuration changed
- added logic to not view keyboard when a user canceled keyboard by tapping back button or 'search' button
- added logic to ignore change of focus and keyboard during configuration changes

Fixes #1033 

**Considerations**
- Need to cache focus flag value during initialization to avoid incorrect modification during initial composition
- Need to ignore focus changes while the activity is changing configurations
- Need to work well with multiple `TextField`s

**Do tests pass?**
- [x] Run local tests on `DemoDebug` variant: `./gradlew testDemoDebug`
- [x] Check formatting: `./gradlew --init-script gradle/init.gradle.kts spotlessApply`

**Is this your first pull request?**
- [x] [Sign the CLA](https://cla.developers.google.com/)
- [x] Run `./tools/setup.sh`
- [x] Import the code formatting style as explained in [the setup script](/tools/setup.sh#L40).

### CODE REVIEW

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+private fun Context.findActivity(): Activity {
+    var context = this
+    while (context is ContextWrapper) {
+        if (context is Activity) return context
+        context = context.baseContext
+    }
+    throw IllegalStateException("Failed to find Activity!")
+}
+
+// Obtain IME visibility as a State which can trigger recomposition.
```
Comment (JoseAlcerreca) [inline feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:53]: Use kdoc format for comments

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+// Note that this may not work if edge-to-edge layout is not enabled.
+@Composable
+private fun imeVisibilityAsState(): State<Boolean> {
+    val isImeVisible = WindowInsets.ime.getBottom(LocalDensity.current) > 0
+    return rememberUpdatedState(isImeVisible)
+}
+
+// Add this modifier to a [TextField] to retain focus and keyboard visibility
+// after configuration is changed. This also works with multiple [TextField]
```
Comment (JoseAlcerreca) [inline feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:62]:
```suggestion
// after configuration is changed. This also works with multiple [TextField]s.
```

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt (hunk 1 above)
Comment (saryongkang) [reply feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:53]: Oh, gotcha. I will correct all comments.

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt (hunk 2 above)
Comment (saryongkang) [reply feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:62]: Ooops. Thanks!

**Comment by saryongkang** [comment]
As per advice from @JolandaVerhoef, before merging this PR, I will wait for Compose Engineers' opinion on this workaround.

**Comment by saryongkang** [comment]
I created another repository to highlight this workaround.

https://github.com/saryongkang/focus-restorable-textfield
Uploading screenrecording_focus_textfield_1113_2023.mp4…

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    val imeVisibilityState by imeVisibilityAsState()
+    LaunchedEffect(Unit) {
+        if (shouldRequestFocus) {
+            focusRequester.requestFocus()
+            shouldRequestFocus = false
+        }
+        snapshotFlow { imeVisibilityState }
+            .drop(2) // Drop the first two initial values when launched
+            .collectLatest { isVisible ->
```
Comment (hoangchungk53qx1) [inline feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:97]: Just collect is enough

**Comment by Fando** [comment]
There are a few issues:

1. The cursor position is not retained after a configuration change
2. IME state is not retained after configuration changes in all cases. Try changing configuration 5 times in a row, and notice the inconsistency. Sometimes IME state would be preserved and sometimes not.

Tested on a physical device, Pixel 6a, Android 13

File: feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.drop
+
+private fun Context.findActivity(): Activity {
+    var context = this
+    while (context is ContextWrapper) {
+        if (context is Activity) return context
+        context = context.baseContext
+    }
+    throw IllegalStateException("Failed to find Activity!")
```
Comment (Fando) [inline feature/search/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/search/FocusRestorerModifier.kt:50]: Maybe worth while to make the return type nullable to allow rendering compose previews.

**Comment by saryongkang** [comment]
Sorry for the delayed process.
Due to several changes in `BasicTextField` and focus management, this PR need to be rewritten to reflect the latest update. Let me close it temporarily and re-open it in Jan 2025.

**Pull request closed without merging**